src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+    }
+    filestats->st_size = file->size;
+    filestats->st_blksize = F_PART_DATA_SIZE;
+    filestats->st_blocks = 1 + file->size % F_PART_DATA_SIZE;
+    return OS_FS_SUCCESS;
+}
+
+int32 file_lseek(int32  filedes, int32 offset, uint32 whence)
+{
+    int32 ret = chk_fd(filedes);
+    if (ret) return ret;
+
+    struct fsobj *o = &files[filedes - 1];
+    uint32 target_offset = 0;
+
+    // wasnt sure if it should be legal to pass negative offset, went with yes
+    if (whence == SEEK_SET) {
+
+        if (offset < 0) return OS_FS_ERROR;
+        target_offset = offset;
+
+    } else if (whence == SEEK_CUR) {
+
+        if (offset + (int32) o->position.file_offset < 0) return OS_FS_ERROR;
+        target_offset += o->position.file_offset;
+
+    } else if (whence == SEEK_END) {
+
+        if (offset + (int32) o->position.file_offset < 0) return OS_FS_ERROR;
+        target_offset += o->size;
+
+    } else {
+        return OS_FS_ERROR;
+    }
+    // you cannot write past the end of a static file
+    if ( target_offset > o->size && o->file_part->memtype == STATIC) {
+        return OS_FS_ERROR;
+    }
+
+    o->position.open_part = o->file_part;
+    o->position.file_offset = 0;
+
+    while ( target_offset - o->position.file_offset > F_PART_DATA_SIZE) {
+
+        // seeking past the end of a file writes zeros until that position
+        if (o->position.open_part->next == NULL) {
+            struct f_part *part;
+            newpart_get(&part);
+            part->memtype               = DYNAMIC;
+            part->file                  = o;
+            part->next                  = NULL;
+            part->prev                  = o->position.open_part;
+            o->position.open_part->next = part;
+        }
+
+        o->position.open_part = o->position.open_part->next;
+        o->position.file_offset += F_PART_DATA_SIZE;
+
+    }
+
+    o->position.file_offset += target_offset % F_PART_DATA_SIZE;
+    o->position.part_offset = target_offset % F_PART_DATA_SIZE;
+    o->position.addr = o->position.open_part->data + target_offset % F_PART_DATA_SIZE;
+
+    if (o->position.file_offset > size) {
+        size = o->position.file_offset;
+    }
+
+    assert(o->position.file_offset == target_offset);
+    return target_offset;
+}
+
+int32 file_cp(const char *src, const char *dest)
+{
+    struct fsobj *file_src, *file_dest;
+    int32 src_desc = file_open(src);
+    if (src_desc <= 0 || src_desc > MAX_NUM_FILES) return OS_FS_ERROR;
+    file_src = &files[src_desc - 1];
+
+    newfile_get(&file_dest);
+
+    if(file_src == NULL || file_src->type == FSOBJ_DIR){
+        return OS_FS_ERR_INVALID_POINTER;
+    }
+
+    file_insert(file_dest, (char *) dest);
+    file_dest->name = path_to_name( (char *) dest);
+    file_dest->size = 0;
+    file_dest->mode = file_src->mode;
+    file_dest->type = FSOBJ_FILE;
+
+    /*
+     * yeah, this is sub-optimal
+     * but I don't expect this to be called at all, much less with large files
+     */
```

Comment:
Could be a little less informal in this comment. There really isn't a much better way to do this.